src/bgw/job.c

```diff
@@ -755,21 +755,24 @@ bgw_job_tuple_update_by_id(TupleInfo *ti, void *const data)
 	BgwJob *updated_job = (BgwJob *) data;
 	HeapTuple tuple = heap_copytuple(ti->tuple);
 	FormData_bgw_job *fd = (FormData_bgw_job *) GETSTRUCT(tuple);
-	TimestampTz nowts, next_start;
+	TimestampTz next_start;
 
 	ts_bgw_job_permission_check(updated_job);
 	/* when we update the schedule interval, modify the next start time as well*/
 	if (!DatumGetBool(DirectFunctionCall2(interval_eq,
 										  IntervalPGetDatum(&fd->schedule_interval),
 										  IntervalPGetDatum(&updated_job->fd.schedule_interval))))
 	{
-		nowts = GetCurrentTimestamp();
+		BgwJobStat *stat = ts_bgw_job_stat_find(fd->id);
 
-		next_start = DatumGetTimestamp(
-			DirectFunctionCall2(timestamp_pl_interval,
-								TimestampGetDatum(nowts),
-								IntervalPGetDatum(&updated_job->fd.schedule_interval)));
-		ts_bgw_job_stat_update_next_start(updated_job, next_start);
+		if (stat != NULL)
+		{
+			next_start = DatumGetTimestampTz(
+				DirectFunctionCall2(timestamptz_pl_interval,
+									TimestampTzGetDatum(stat->fd.last_finish),
+									IntervalPGetDatum(&updated_job->fd.schedule_interval)));
+			ts_bgw_job_stat_update_next_start(updated_job, next_start);
```

Comment:
Should we have a default next-start-time if there is no previous `stat`?

Reply:
If there is no stat, the scheduler already interprets as -infinity which I think is the right thing